// @zag-js/drawer/dist/index.js
'use strict';

var anatomy$1 = require('@zag-js/anatomy');
var domQuery = require('@zag-js/dom-query');
var utils = require('@zag-js/utils');
var ariaHidden = require('@zag-js/aria-hidden');
var core = require('@zag-js/core');
var dismissable = require('@zag-js/dismissable');
var focusTrap = require('@zag-js/focus-trap');
var removeScroll = require('@zag-js/remove-scroll');
var types = require('@zag-js/types');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var anatomy = anatomy$1.createAnatomy("drawer").parts(
  "positioner",
  "content",
  "title",
  "trigger",
  "backdrop",
  "grabber",
  "grabberIndicator",
  "closeTrigger"
);
var parts = anatomy.build();
var getContentId = (ctx) => ctx.ids?.content ?? `drawer:${ctx.id}:content`;
var getPositionerId = (ctx) => ctx.ids?.positioner ?? `drawer:${ctx.id}:positioner`;
var getTitleId = (ctx) => ctx.ids?.title ?? `drawer:${ctx.id}:title`;
var getTriggerId = (ctx) => ctx.ids?.trigger ?? `drawer:${ctx.id}:trigger`;
var getBackdropId = (ctx) => ctx.ids?.backdrop ?? `drawer:${ctx.id}:backdrop`;
var getGrabberId = (ctx) => ctx.ids?.grabber ?? `drawer:${ctx.id}:grabber`;
var getGrabberIndicatorId = (ctx) => ctx.ids?.grabberIndicator ?? `drawer:${ctx.id}:grabber-indicator`;
var getCloseTriggerId = (ctx) => ctx.ids?.closeTrigger ?? `drawer:${ctx.id}:close-trigger`;
var getContentEl = (ctx) => ctx.getById(getContentId(ctx));
var getTriggerEl = (ctx) => ctx.getById(getTriggerId(ctx));
var getCloseTriggerEl = (ctx) => ctx.getById(getCloseTriggerId(ctx));

// src/drawer.connect.ts
var isVerticalDirection = (direction) => direction === "down" || direction === "up";
var isNegativeDirection = (direction) => direction === "up" || direction === "left";
function connect(service, normalize) {
  const { state, send, context, scope, prop } = service;
  const open = state.hasTag("open");
  const closed = state.matches("closed");
  const dragOffset = context.get("dragOffset");
  const dragging = dragOffset !== null;
  const snapPoint = context.get("snapPoint");
  const resolvedActiveSnapPoint = context.get("resolvedActiveSnapPoint");
  const swipeDirection = prop("swipeDirection");
  const contentSize = context.get("contentSize");
  const swipeStrength = context.get("swipeStrength");
  const snapPointOffset = resolvedActiveSnapPoint?.offset ?? 0;
  const currentOffset = dragOffset ?? snapPointOffset;
  const swipeMovement = dragging ? currentOffset - snapPointOffset : 0;
  const signedCurrentOffset = isNegativeDirection(swipeDirection) ? -currentOffset : currentOffset;
  const signedSnapPointOffset = isNegativeDirection(swipeDirection) ? -snapPointOffset : snapPointOffset;
  const signedMovement = isNegativeDirection(swipeDirection) ? -swipeMovement : swipeMovement;
  const swipeProgress = dragging && contentSize && contentSize > 0 ? Math.max(0, Math.min(1, Math.abs(signedMovement) / contentSize)) : 0;
  const translateX = isVerticalDirection(swipeDirection) ? 0 : signedCurrentOffset;
  const translateY = isVerticalDirection(swipeDirection) ? signedCurrentOffset : 0;
  function onPointerDown(event) {
    if (!domQuery.isLeftClick(event)) return;
    const target = domQuery.getEventTarget(event);
    if (target?.hasAttribute("data-no-drag") || target?.closest("[data-no-drag]")) return;
    if (state.matches("closing")) return;
    send({ type: "POINTER_DOWN", point: domQuery.getEventPoint(event) });
  }
  return {
    open,
    dragging,
    setOpen(nextOpen) {
      const open2 = state.hasTag("open");
      if (open2 === nextOpen) return;
      send({ type: nextOpen ? "OPEN" : "CLOSE" });
    },
    snapPoints: prop("snapPoints"),
    swipeDirection,
    snapPoint,
    setSnapPoint(snapPoint2) {
      const currentSnapPoint = context.get("snapPoint");
      if (currentSnapPoint === snapPoint2) return;
      send({ type: "SNAP_POINT.SET", snapPoint: snapPoint2 });
    },
    getOpenPercentage() {
      if (!open) return 0;
      if (!contentSize) return 0;
      return Math.max(0, Math.min(1, 1 - currentOffset / contentSize));
    },
    getSnapPointIndex() {
      const snapPoints = prop("snapPoints");
      if (snapPoint === null) return -1;
      return snapPoints.indexOf(snapPoint);
    },
    getContentSize() {
      return contentSize;
    },
    getPositionerProps() {
      return normalize.element({
        ...parts.positioner.attrs,
        id: getPositionerId(scope),
        dir: prop("dir"),
        hidden: closed,
        "data-state": open ? "open" : "closed",
        "data-swipe-direction": swipeDirection
      });
    },
    getContentProps(props2 = { draggable: true }) {
      const movementX = isVerticalDirection(swipeDirection) ? 0 : signedMovement;
      const movementY = isVerticalDirection(swipeDirection) ? signedMovement : 0;
      return normalize.element({
        ...parts.content.attrs,
        dir: prop("dir"),
        id: getContentId(scope),
        tabIndex: -1,
        role: prop("role"),
        "aria-modal": prop("modal"),
        "aria-labelledby": getTitleId(scope),
        hidden: !open,
        "data-state": open ? "open" : "closed",
        "data-expanded": resolvedActiveSnapPoint?.offset === 0 ? "" : void 0,
        "data-swipe-direction": swipeDirection,
        "data-swiping": dragging ? "" : void 0,
        "data-dragging": dragging ? "" : void 0,
        style: {
          transform: "translate3d(var(--drawer-translate-x, 0px), var(--drawer-translate-y, 0px), 0)",
          transitionDuration: dragging ? "0s" : void 0,
          "--drawer-translate": utils.toPx(translateY),
          "--drawer-translate-x": utils.toPx(translateX),
          "--drawer-translate-y": utils.toPx(translateY),
          "--drawer-snap-point-offset-x": isVerticalDirection(swipeDirection) ? "0px" : utils.toPx(signedSnapPointOffset),
          "--drawer-snap-point-offset-y": isVerticalDirection(swipeDirection) ? utils.toPx(signedSnapPointOffset) : "0px",
          "--drawer-swipe-movement-x": utils.toPx(movementX),
          "--drawer-swipe-movement-y": utils.toPx(movementY),
          "--drawer-swipe-strength": `${swipeStrength}`,
          willChange: "transform"
        },
        onPointerDown(event) {
          if (!props2.draggable) return;
          onPointerDown(event);
        }
      });
    },
    getTitleProps() {
      return normalize.element({
        ...parts.title.attrs,
        id: getTitleId(scope),
        dir: prop("dir")
      });
    },
    getTriggerProps() {
      return normalize.button({
        ...parts.trigger.attrs,
        id: getTriggerId(scope),
        type: "button",
        onClick() {
          send({ type: open ? "CLOSE" : "OPEN" });
        }
      });
    },
    getBackdropProps() {
      return normalize.element({
        ...parts.backdrop.attrs,
        id: getBackdropId(scope),
        hidden: !open,
        "data-state": open ? "open" : "closed",
        "data-swiping": dragging ? "" : void 0,
        style: {
          willChange: "opacity",
          "--drawer-swipe-progress": `${swipeProgress}`,
          "--drawer-swipe-strength": `${swipeStrength}`
        }
      });
    },
    getGrabberProps() {
      return normalize.element({
        ...parts.grabber.attrs,
        id: getGrabberId(scope),
        onPointerDown(event) {
          onPointerDown(event);
        },
        style: {
          touchAction: "none"
        }
      });
    },
    getGrabberIndicatorProps() {
      return normalize.element({
        ...parts.grabberIndicator.attrs,
        id: getGrabberIndicatorId(scope)
      });
    },
    getCloseTriggerProps() {
      return normalize.button({
        ...parts.closeTrigger.attrs,
        id: getCloseTriggerId(scope),
        onClick() {
          send({ type: "CLOSE" });
        }
      });
    }
  };
}

// src/drawer.registry.ts
var elements = /* @__PURE__ */ new Map();
var swipingIds = /* @__PURE__ */ new Set();
var listeners = /* @__PURE__ */ new Set();
function notify() {
  listeners.forEach((fn) => fn());
}
var drawerRegistry = {
  register(id, el) {
    elements.set(id, el);
    notify();
  },
  unregister(id) {
    swipingIds.delete(id);
    if (!elements.delete(id)) return;
    notify();
  },
  setSwiping(id, swiping) {
    const changed = swiping ? !swipingIds.has(id) : swipingIds.has(id);
    if (!changed) return;
    if (swiping) swipingIds.add(id);
    else swipingIds.delete(id);
    notify();
  },
  hasSwipingAfter(id) {
    const keys = [...elements.keys()];
    const myIndex = keys.indexOf(id);
    if (myIndex === -1) return false;
    return keys.slice(myIndex + 1).some((key) => swipingIds.has(key));
  },
  notify,
  getEntries() {
    return elements;
  },
  subscribe(fn) {
    listeners.add(fn);
    return () => {
      listeners.delete(fn);
    };
  }
};

// src/utils/find-closest-snap-point.ts
function findClosestSnapPoint(offset, snapPoints) {
  return snapPoints.reduce((acc, curr) => {
    const closestDiff = Math.abs(offset - acc.offset);
    const currentDiff = Math.abs(offset - curr.offset);
    return currentDiff < closestDiff ? curr : acc;
  });
}

// src/utils/get-scroll-info.ts
function isVerticalScrollContainer(element) {
  const styles = getComputedStyle(element);
  const overflow = styles.overflowY;
  return overflow === "auto" || overflow === "scroll";
}
function isHorizontalScrollContainer(element) {
  const styles = getComputedStyle(element);
  const overflow = styles.overflowX;
  return overflow === "auto" || overflow === "scroll";
}
var isVerticalDirection2 = (direction) => direction === "up" || direction === "down";
function getScrollInfo(target, container, direction) {
  let element = target;
  let availableForwardScroll = 0;
  let availableBackwardScroll = 0;
  const vertical = isVerticalDirection2(direction);
  while (element) {
    const clientSize = vertical ? element.clientHeight : element.clientWidth;
    const scrollPos = vertical ? element.scrollTop : element.scrollLeft;
    const scrollSize = vertical ? element.scrollHeight : element.scrollWidth;
    const scrolled = scrollSize - scrollPos - clientSize;
    const isScrollable = vertical ? isVerticalScrollContainer(element) : isHorizontalScrollContainer(element);
    if ((scrollPos !== 0 || scrolled !== 0) && isScrollable) {
      availableForwardScroll += scrolled;
      availableBackwardScroll += scrollPos;
    }
    if (element === container || element === document.documentElement) break;
    element = element.parentNode;
  }
  return {
    availableForwardScroll,
    availableBackwardScroll
  };
}

// src/utils/drag-manager.ts
var DRAG_START_THRESHOLD = 0.3;
var SEQUENTIAL_THRESHOLD = 14;
var SNAP_VELOCITY_THRESHOLD = 500;
var SNAP_VELOCITY_MULTIPLIER = 0.3;
var MAX_SNAP_VELOCITY = 4e3;
var DragManager = class {
  constructor() {
    __publicField(this, "pointerStart", null);
    __publicField(this, "dragOffset", null);
    __publicField(this, "lastPoint", null);
    __publicField(this, "lastTimestamp", null);
    __publicField(this, "velocity", null);
  }
  setPointerStart(point) {
    this.pointerStart = point;
  }
  clearPointerStart() {
    this.pointerStart = null;
  }
  getPointerStart() {
    return this.pointerStart;
  }
  getAxisValue(point, direction) {
    return direction === "left" || direction === "right" ? point.x : point.y;
  }
  getDirectionSign(direction) {
    return direction === "up" || direction === "left" ? -1 : 1;
  }
  setDragOffsetForDirection(point, resolvedActiveSnapPointOffset, direction) {
    if (!this.pointerStart) return;
    const currentTimestamp = (/* @__PURE__ */ new Date()).getTime();
    const sign = this.getDirectionSign(direction);
    const axisValue = this.getAxisValue(point, direction);
    if (this.lastPoint) {
      const lastAxisValue = this.getAxisValue(this.lastPoint, direction);
      const delta2 = (axisValue - lastAxisValue) * sign;
      if (this.lastTimestamp) {
        const dt = currentTimestamp - this.lastTimestamp;
        if (dt > 0) {
          const calculatedVelocity = delta2 / dt * 1e3;
          this.velocity = Number.isFinite(calculatedVelocity) ? calculatedVelocity : 0;
        }
      }
    }
    this.lastPoint = point;
    this.lastTimestamp = currentTimestamp;
    const pointerStartAxis = this.getAxisValue(this.pointerStart, direction);
    let delta = (pointerStartAxis - axisValue) * sign - resolvedActiveSnapPointOffset;
    if (delta > 0) {
      delta = Math.sqrt(delta);
    }
    this.dragOffset = -delta;
  }
  getDragOffset() {
    return this.dragOffset;
  }
  clearDragOffset() {
    this.dragOffset = null;
  }
  getVelocity() {
    return this.velocity;
  }
  clearVelocityTracking() {
    this.lastPoint = null;
    this.lastTimestamp = null;
    this.velocity = null;
  }
  clear() {
    this.clearPointerStart();
    this.clearDragOffset();
    this.clearVelocityTracking();
  }
  shouldStartDragging(point, target, container, preventDragOnScroll, direction) {
    if (!this.pointerStart || !container) return false;
    if (preventDragOnScroll) {
      const sign = this.getDirectionSign(direction);
      const axis = this.getAxisValue(point, direction);
      const startAxis = this.getAxisValue(this.pointerStart, direction);
      const delta = (startAxis - axis) * sign;
      if (Math.abs(delta) < DRAG_START_THRESHOLD) return false;
      const { availableForwardScroll, availableBackwardScroll } = getScrollInfo(target, container, direction);
      if (delta > 0 && Math.abs(availableForwardScroll) > 1 || delta < 0 && Math.abs(availableBackwardScroll) > 0) {
        return false;
      }
    }
    return true;
  }
  findClosestSnapPoint(snapPoints, snapPoint, snapToSequentialPoints) {
    if (this.dragOffset === null) {
      return snapPoints[0]?.value ?? 1;
    }
    if (snapToSequentialPoints && snapPoint) {
      const currentIndex = snapPoints.findIndex((item) => Object.is(item.value, snapPoint.value));
      if (currentIndex >= 0) {
        const delta = this.dragOffset - snapPoint.offset;
        const dragDirection = Math.sign(delta);
        const velocityDirection = this.velocity !== null ? Math.sign(this.velocity) : 0;
        const shouldAdvance = dragDirection !== 0 && velocityDirection === dragDirection && Math.abs(this.velocity ?? 0) >= SNAP_VELOCITY_THRESHOLD;
        if (shouldAdvance) {
          const nextIndex = Math.min(Math.max(currentIndex + dragDirection, 0), snapPoints.length - 1);
          if (nextIndex !== currentIndex) {
            return snapPoints[nextIndex].value;
          }
        }
        if (delta > SEQUENTIAL_THRESHOLD) {
          return snapPoints[Math.min(currentIndex + 1, snapPoints.length - 1)]?.value ?? snapPoint.value;
        }
        if (delta < -SEQUENTIAL_THRESHOLD) {
          return snapPoints[Math.max(currentIndex - 1, 0)]?.value ?? snapPoint.value;
        }
        return snapPoint.value;
      }
    }
    let targetOffset = this.dragOffset;
    if (this.velocity !== null && Math.abs(this.velocity) >= SNAP_VELOCITY_THRESHOLD) {
      const clamped = Math.min(MAX_SNAP_VELOCITY, Math.max(-MAX_SNAP_VELOCITY, this.velocity));
      targetOffset += clamped * SNAP_VELOCITY_MULTIPLIER;
      targetOffset = Math.max(0, targetOffset);
    }
    const closest = findClosestSnapPoint(targetOffset, snapPoints);
    return closest.value;
  }
  computeSwipeStrength(targetOffset) {
    const MAX_DURATION_MS = 360;
    const MIN_SCALAR = 0.1;
    const MAX_SCALAR = 1;
    if (this.dragOffset === null || this.velocity === null) return MAX_SCALAR;
    const distance = Math.abs(this.dragOffset - targetOffset);
    const absVelocity = Math.abs(this.velocity);
    if (absVelocity <= 0 || distance <= 0) return MAX_SCALAR;
    const estimatedTimeMs = distance / absVelocity * 1e3;
    const normalized = Math.min(1, Math.max(0, estimatedTimeMs / MAX_DURATION_MS));
    return MIN_SCALAR + normalized * (MAX_SCALAR - MIN_SCALAR);
  }
  shouldDismiss(contentSize, snapPoints, swipeVelocityThreshold, closeThreshold) {
    if (this.dragOffset === null || this.velocity === null || contentSize === null) return false;
    const visibleSize = contentSize - this.dragOffset;
    const smallestSnapPoint = snapPoints.reduce((acc, curr) => curr.offset > acc.offset ? curr : acc);
    const isFastSwipe = this.velocity > 0 && this.velocity >= swipeVelocityThreshold;
    const closeThresholdInPixels = contentSize * (1 - closeThreshold);
    const smallestSnapPointVisibleSize = contentSize - smallestSnapPoint.offset;
    const isBelowSmallestSnapPoint = visibleSize < smallestSnapPointVisibleSize;
    const isBelowCloseThreshold = visibleSize < closeThresholdInPixels;
    if (snapPoints.length > 1) {
      return visibleSize <= 0 || isFastSwipe && isBelowSmallestSnapPoint;
    }
    const hasEnoughDragToDismiss = isBelowCloseThreshold && isBelowSmallestSnapPoint || visibleSize === 0;
    return isFastSwipe && isBelowSmallestSnapPoint || hasEnoughDragToDismiss;
  }
};

// src/utils/resolve-snap-point.ts
function clamp(value, min, max) {
  return Math.min(Math.max(value, min), max);
}
function resolveSnapPointValue(snapPoint, viewportSize, rootFontSize) {
  if (!Number.isFinite(viewportSize) || viewportSize <= 0) return null;
  if (typeof snapPoint === "number") {
    if (!Number.isFinite(snapPoint)) return null;
    if (snapPoint <= 1) return clamp(snapPoint, 0, 1) * viewportSize;
    return snapPoint;
  }
  const trimmed = snapPoint.trim();
  if (trimmed.endsWith("px")) {
    const value = Number.parseFloat(trimmed);
    return Number.isFinite(value) ? value : null;
  }
  if (trimmed.endsWith("rem")) {
    const value = Number.parseFloat(trimmed);
    return Number.isFinite(value) ? value * rootFontSize : null;
  }
  return null;
}
function resolveSnapPoint(snapPoint, options) {
  const { popupSize, viewportSize, rootFontSize } = options;
  const maxSize = Math.min(popupSize, viewportSize);
  if (!Number.isFinite(maxSize) || maxSize <= 0) return null;
  const resolvedSize = resolveSnapPointValue(snapPoint, viewportSize, rootFontSize);
  if (resolvedSize === null || !Number.isFinite(resolvedSize)) return null;
  const height = clamp(resolvedSize, 0, maxSize);
  return {
    value: snapPoint,
    height,
    offset: Math.max(0, popupSize - height)
  };
}

// src/drawer.machine.ts
var isVerticalDirection3 = (direction) => direction === "down" || direction === "up";
function dedupeSnapPoints(points) {
  if (points.length <= 1) return points;
  const deduped = [];
  const seenHeights = [];
  for (let index = points.length - 1; index >= 0; index -= 1) {
    const point = points[index];
    const isDuplicate = seenHeights.some((height) => Math.abs(height - point.height) <= 1);
    if (isDuplicate) continue;
    seenHeights.push(point.height);
    deduped.push(point);
  }
  deduped.reverse();
  return deduped;
}
function getDirectionSize(rect, direction) {
  return isVerticalDirection3(direction) ? rect.height : rect.width;
}
function clamp2(value, min, max) {
  return Math.min(max, Math.max(min, value));
}
function resolveSwipeProgress(contentSize, dragOffset, snapPointOffset) {
  if (!contentSize || contentSize <= 0) return 0;
  const currentOffset = dragOffset ?? snapPointOffset;
  return clamp2(1 - currentOffset / contentSize, 0, 1);
}
var machine = core.createMachine({
  props({ props: props2, scope }) {
    const alertDialog = props2.role === "alertdialog";
    const initialFocusEl = alertDialog ? () => getCloseTriggerEl(scope) : void 0;
    const modal = typeof props2.modal === "boolean" ? props2.modal : true;
    const snapPoints = props2.snapPoints ?? [1];
    return {
      modal,
      trapFocus: modal,
      preventScroll: modal,
      closeOnInteractOutside: true,
      closeOnEscape: true,
      restoreFocus: true,
      role: "dialog",
      initialFocusEl,
      snapPoints,
      defaultSnapPoint: props2.defaultSnapPoint ?? snapPoints[0] ?? null,
      swipeDirection: "down",
      snapToSequentialPoints: false,
      swipeVelocityThreshold: 700,
      closeThreshold: 0.25,
      preventDragOnScroll: true,
      ...props2
    };
  },
  context({ bindable, prop }) {
    return {
      dragOffset: bindable(() => ({
        defaultValue: null
      })),
      snapPoint: bindable(() => ({
        defaultValue: prop("defaultSnapPoint"),
        value: prop("snapPoint"),
        onChange(value) {
          return prop("onSnapPointChange")?.({ snapPoint: value });
        }
      })),
      resolvedActiveSnapPoint: bindable(() => ({
        defaultValue: null
      })),
      contentSize: bindable(() => ({
        defaultValue: null
      })),
      viewportSize: bindable(() => ({
        defaultValue: 0
      })),
      rootFontSize: bindable(() => ({
        defaultValue: 16
      })),
      swipeStrength: bindable(() => ({
        defaultValue: 1
      }))
    };
  },
  refs() {
    return {
      dragManager: new DragManager()
    };
  },
  computed: {
    resolvedSnapPoints({ context, prop }) {
      const contentSize = context.get("contentSize");
      const viewportSize = context.get("viewportSize");
      const rootFontSize = context.get("rootFontSize");
      if (contentSize === null) return [];
      const points = prop("snapPoints").map(
        (snapPoint) => resolveSnapPoint(snapPoint, {
          popupSize: contentSize,
          viewportSize,
          rootFontSize
        })
      ).filter((point) => point !== null);
      return dedupeSnapPoints(points);
    }
  },
  watch({ track, context, prop, action, computed }) {
    track(
      [
        () => context.get("snapPoint"),
        () => context.get("contentSize"),
        () => context.get("viewportSize"),
        () => context.get("rootFontSize"),
        () => prop("snapPoints").join("|")
      ],
      () => {
        const snapPoint = context.get("snapPoint");
        const contentSize = context.get("contentSize");
        const viewportSize = context.get("viewportSize");
        const rootFontSize = context.get("rootFontSize");
        if (snapPoint === null || contentSize === null) {
          context.set("resolvedActiveSnapPoint", null);
          return;
        }
        const resolvedPoints = computed("resolvedSnapPoints");
        const matchedPoint = resolvedPoints.find((point) => Object.is(point.value, snapPoint));
        if (matchedPoint) {
          context.set("resolvedActiveSnapPoint", matchedPoint);
          return;
        }
        const resolvedActiveSnapPoint = resolveSnapPoint(snapPoint, {
          popupSize: contentSize,
          viewportSize,
          rootFontSize
        });
        if (resolvedActiveSnapPoint) {
          context.set("resolvedActiveSnapPoint", resolvedActiveSnapPoint);
          return;
        }
        const fallbackPoint = resolvedPoints[0];
        if (!fallbackPoint) {
          context.set("resolvedActiveSnapPoint", null);
          return;
        }
        context.set("snapPoint", fallbackPoint.value);
        context.set("resolvedActiveSnapPoint", fallbackPoint);
      }
    );
    track([() => prop("open")], () => {
      action(["toggleVisibility"]);
    });
    track(
      [
        () => context.get("dragOffset"),
        () => context.get("contentSize"),
        () => context.get("resolvedActiveSnapPoint")?.offset ?? 0
      ],
      () => {
        action(["syncDrawerStack"]);
      }
    );
  },
  initialState({ prop }) {
    const open = prop("open") || prop("defaultOpen");
    return open ? "open" : "closed";
  },
  on: {
    "SNAP_POINT.SET": {
      actions: ["setSnapPoint"]
    }
  },
  states: {
    open: {
      tags: ["open"],
      effects: [
        "trackDismissableElement",
        "preventScroll",
        "trapFocus",
        "hideContentBelow",
        "trackPointerMove",
        "trackSizeMeasurements",
        "trackNestedDrawerMetrics",
        "trackDrawerStack"
      ],
      on: {
        "CONTROLLED.CLOSE": {
          target: "closing",
          actions: ["resetSwipeStrength"]
        },
        POINTER_DOWN: {
          actions: ["setPointerStart"]
        },
        POINTER_MOVE: [
          {
            guard: "isDragging",
            actions: ["setDragOffset"]
          },
          {
            guard: "shouldStartDragging",
            actions: ["setRegistrySwiping", "setDragOffset"]
          }
        ],
        POINTER_UP: [
          {
            guard: "shouldCloseOnSwipe",
            target: "closing",
            actions: ["clearRegistrySwiping", "setDismissSwipeStrength"]
          },
          {
            guard: "isDragging",
            actions: [
              "clearRegistrySwiping",
              "setSnapSwipeStrength",
              "setClosestSnapPoint",
              "clearPointerStart",
              "clearDragOffset"
            ]
          },
          {
            actions: ["clearRegistrySwiping", "clearPointerStart", "clearDragOffset"]
          }
        ],
        CLOSE: [
          {
            guard: "isOpenControlled",
            actions: ["invokeOnClose"]
          },
          {
            target: "closing",
            actions: ["resetSwipeStrength", "invokeOnClose"]
          }
        ]
      }
    },
    closing: {
      effects: ["trackExitAnimation", "trackDrawerStack"],
      on: {
        ANIMATION_END: {
          target: "closed",
          actions: [
            "invokeOnClose",
            "clearPointerStart",
            "clearDragOffset",
            "clearActiveSnapPoint",
            "clearResolvedActiveSnapPoint",
            "clearSizeMeasurements",
            "clearVelocityTracking"
          ]
        }
      }
    },
    closed: {
      tags: ["closed"],
      on: {
        "CONTROLLED.OPEN": {
          target: "open"
        },
        OPEN: [
          {
            guard: "isOpenControlled",
            actions: ["invokeOnOpen"]
          },
          {
            target: "open",
            actions: ["invokeOnOpen"]
          }
        ]
      }
    }
  },
  implementations: {
    guards: {
      isOpenControlled: ({ prop }) => prop("open") !== void 0,
      isDragging({ context }) {
        return context.get("dragOffset") !== null;
      },
      shouldStartDragging({ prop, refs, event, scope }) {
        if (!(event.target instanceof HTMLElement)) return false;
        const dragManager = refs.get("dragManager");
        return dragManager.shouldStartDragging(
          event.point,
          event.target,
          getContentEl(scope),
          prop("preventDragOnScroll"),
          prop("swipeDirection")
        );
      },
      shouldCloseOnSwipe({ prop, context, computed, refs }) {
        const dragManager = refs.get("dragManager");
        return dragManager.shouldDismiss(
          context.get("contentSize"),
          computed("resolvedSnapPoints"),
          prop("swipeVelocityThreshold"),
          prop("closeThreshold")
        );
      }
    },
    actions: {
      invokeOnOpen({ prop }) {
        prop("onOpenChange")?.({ open: true });
      },
      invokeOnClose({ prop }) {
        prop("onOpenChange")?.({ open: false });
      },
      setSnapPoint({ context, event }) {
        context.set("snapPoint", event.snapPoint);
      },
      setPointerStart({ event, refs }) {
        refs.get("dragManager").setPointerStart(event.point);
      },
      setDragOffset({ context, event, refs, prop }) {
        const dragManager = refs.get("dragManager");
        dragManager.setDragOffsetForDirection(
          event.point,
          context.get("resolvedActiveSnapPoint")?.offset || 0,
          event.swipeDirection ?? prop("swipeDirection")
        );
        context.set("dragOffset", dragManager.getDragOffset());
      },
      setClosestSnapPoint({ computed, context, refs, prop }) {
        const snapPoints = computed("resolvedSnapPoints");
        const contentSize = context.get("contentSize");
        const viewportSize = context.get("viewportSize");
        const rootFontSize = context.get("rootFontSize");
        if (!snapPoints.length || contentSize === null) return;
        const dragManager = refs.get("dragManager");
        const closestSnapPoint = dragManager.findClosestSnapPoint(
          snapPoints,
          context.get("resolvedActiveSnapPoint"),
          prop("snapToSequentialPoints")
        );
        context.set("snapPoint", closestSnapPoint);
        const resolved = resolveSnapPoint(closestSnapPoint, {
          popupSize: contentSize,
          viewportSize,
          rootFontSize
        });
        context.set("resolvedActiveSnapPoint", resolved);
      },
      clearDragOffset({ context, refs }) {
        refs.get("dragManager").clearDragOffset();
        context.set("dragOffset", null);
      },
      clearActiveSnapPoint({ context, prop }) {
        context.set("snapPoint", prop("defaultSnapPoint") ?? null);
      },
      clearSizeMeasurements({ context }) {
        context.set("contentSize", null);
        context.set("viewportSize", 0);
        context.set("rootFontSize", 16);
      },
      clearResolvedActiveSnapPoint({ context }) {
        context.set("resolvedActiveSnapPoint", null);
      },
      clearPointerStart({ refs }) {
        refs.get("dragManager").clearPointerStart();
      },
      clearVelocityTracking({ refs }) {
        refs.get("dragManager").clearVelocityTracking();
      },
      setSnapSwipeStrength({ context, refs, computed, prop }) {
        const dragManager = refs.get("dragManager");
        const snapPoints = computed("resolvedSnapPoints");
        const closestSnapPoint = dragManager.findClosestSnapPoint(
          snapPoints,
          context.get("resolvedActiveSnapPoint"),
          prop("snapToSequentialPoints")
        );
        const contentSize = context.get("contentSize");
        const viewportSize = context.get("viewportSize");
        const rootFontSize = context.get("rootFontSize");
        const resolved = resolveSnapPoint(closestSnapPoint, {
          popupSize: contentSize ?? 0,
          viewportSize,
          rootFontSize
        });
        context.set("swipeStrength", dragManager.computeSwipeStrength(resolved?.offset ?? 0));
      },
      setDismissSwipeStrength({ context, refs }) {
        const dragManager = refs.get("dragManager");
        const contentSize = context.get("contentSize");
        context.set("swipeStrength", dragManager.computeSwipeStrength(contentSize ?? 0));
      },
      resetSwipeStrength({ context }) {
        context.set("swipeStrength", 1);
      },
      setRegistrySwiping({ scope, prop }) {
        const id = String(prop("id") ?? scope.id);
        drawerRegistry.setSwiping(id, true);
      },
      clearRegistrySwiping({ scope, prop }) {
        const id = String(prop("id") ?? scope.id);
        drawerRegistry.setSwiping(id, false);
      },
      toggleVisibility({ event, send, prop }) {
        send({ type: prop("open") ? "CONTROLLED.OPEN" : "CONTROLLED.CLOSE", previousEvent: event });
      },
      syncDrawerStack({ context, prop, scope }) {
        const stack = prop("stack");
        if (!stack) return;
        const contentSize = context.get("contentSize");
        if (contentSize === null) return;
        const id = String(prop("id") ?? scope.id);
        const dragOffset = context.get("dragOffset");
        const snapPointOffset = context.get("resolvedActiveSnapPoint")?.offset ?? 0;
        stack.setHeight(id, contentSize);
        stack.setSwipe(id, dragOffset !== null, resolveSwipeProgress(contentSize, dragOffset, snapPointOffset));
      }
    },
    effects: {
      trackDrawerStack({ context, scope, prop }) {
        const stack = prop("stack");
        if (!stack) return;
        const id = String(prop("id") ?? scope.id);
        stack.register(id);
        stack.setOpen(id, true);
        const sync = () => {
          const contentSize = context.get("contentSize");
          const dragOffset = context.get("dragOffset");
          const snapPointOffset = context.get("resolvedActiveSnapPoint")?.offset ?? 0;
          stack.setHeight(id, contentSize ?? 0);
          stack.setSwipe(id, dragOffset !== null, resolveSwipeProgress(contentSize, dragOffset, snapPointOffset));
        };
        sync();
        return () => {
          stack.setSwipe(id, false, 0);
          stack.setOpen(id, false);
          stack.unregister(id);
        };
      },
      trackDismissableElement({ scope, prop, send }) {
        const getContentEl2 = () => getContentEl(scope);
        return dismissable.trackDismissableElement(getContentEl2, {
          type: "drawer",
          defer: true,
          pointerBlocking: prop("modal"),
          exclude: [getTriggerEl(scope)],
          onInteractOutside(event) {
            prop("onInteractOutside")?.(event);
            if (!prop("closeOnInteractOutside")) {
              event.preventDefault();
            }
          },
          onFocusOutside: prop("onFocusOutside"),
          onEscapeKeyDown(event) {
            prop("onEscapeKeyDown")?.(event);
            if (!prop("closeOnEscape")) {
              event.preventDefault();
            }
          },
          onPointerDownOutside: prop("onPointerDownOutside"),
          onRequestDismiss: prop("onRequestDismiss"),
          onDismiss() {
            send({ type: "CLOSE", src: "interact-outside" });
          }
        });
      },
      preventScroll({ scope, prop }) {
        if (!prop("preventScroll")) return;
        return removeScroll.preventBodyScroll(scope.getDoc());
      },
      trapFocus({ scope, prop }) {
        if (!prop("trapFocus")) return;
        const contentEl = () => getContentEl(scope);
        return focusTrap.trapFocus(contentEl, {
          preventScroll: true,
          returnFocusOnDeactivate: !!prop("restoreFocus"),
          initialFocus: prop("initialFocusEl"),
          setReturnFocus: (el) => prop("finalFocusEl")?.() || el,
          getShadowRoot: true
        });
      },
      hideContentBelow({ scope, prop }) {
        if (!prop("modal")) return;
        const getElements = () => [getContentEl(scope)];
        return ariaHidden.ariaHidden(getElements, { defer: true });
      },
      trackPointerMove({ scope, send, prop }) {
        let lastAxis = 0;
        const swipeDirection = prop("swipeDirection");
        const isVertical = isVerticalDirection3(swipeDirection);
        function onPointerMove(event) {
          const point = domQuery.getEventPoint(event);
          const target = domQuery.getEventTarget(event);
          send({ type: "POINTER_MOVE", point, target, swipeDirection });
        }
        function onPointerUp(event) {
          if (event.pointerType === "touch") return;
          const point = domQuery.getEventPoint(event);
          send({ type: "POINTER_UP", point });
        }
        function onTouchStart(event) {
          if (!event.touches[0]) return;
          lastAxis = isVertical ? event.touches[0].clientY : event.touches[0].clientX;
        }
        function onTouchMove(event) {
          if (!event.touches[0]) return;
          const point = domQuery.getEventPoint(event);
          const target = event.target;
          if (!prop("preventDragOnScroll")) {
            send({ type: "POINTER_MOVE", point, target, swipeDirection });
            return;
          }
          const contentEl = getContentEl(scope);
          if (!contentEl) return;
          let el = target;
          while (el && el !== contentEl && (isVertical ? el.scrollHeight <= el.clientHeight : el.scrollWidth <= el.clientWidth)) {
            el = el.parentElement;
          }
          if (el && el !== contentEl) {
            const scrollPos = isVertical ? el.scrollTop : el.scrollLeft;
            const axis = isVertical ? event.touches[0].clientY : event.touches[0].clientX;
            const atStart = scrollPos <= 0;
            const movingTowardStart = axis > lastAxis;
            if (atStart && movingTowardStart) {
              event.preventDefault();
            }
            lastAxis = axis;
          }
          send({ type: "POINTER_MOVE", point, target, swipeDirection });
        }
        function onTouchEnd(event) {
          if (event.touches.length !== 0) return;
          const point = domQuery.getEventPoint(event);
          send({ type: "POINTER_UP", point });
        }
        const doc = scope.getDoc();
        const cleanups = [
          domQuery.addDomEvent(doc, "pointermove", onPointerMove),
          domQuery.addDomEvent(doc, "pointerup", onPointerUp),
          domQuery.addDomEvent(doc, "touchstart", onTouchStart, { passive: false }),
          domQuery.addDomEvent(doc, "touchmove", onTouchMove, { passive: false }),
          domQuery.addDomEvent(doc, "touchend", onTouchEnd)
        ];
        return () => {
          cleanups.forEach((cleanup) => cleanup());
        };
      },
      trackSizeMeasurements({ context, scope, prop }) {
        const contentEl = getContentEl(scope);
        if (!contentEl) return;
        const doc = scope.getDoc();
        const html = doc.documentElement;
        const updateSize = () => {
          const direction = prop("swipeDirection");
          const rect = contentEl.getBoundingClientRect();
          const viewportSize = isVerticalDirection3(direction) ? html.clientHeight : html.clientWidth;
          const rootFontSize = Number.parseFloat(getComputedStyle(html).fontSize);
          context.set("contentSize", getDirectionSize(rect, direction));
          context.set("viewportSize", viewportSize);
          if (Number.isFinite(rootFontSize)) {
            context.set("rootFontSize", rootFontSize);
          }
        };
        updateSize();
        const cleanups = [
          domQuery.resizeObserverBorderBox.observe(contentEl, updateSize),
          domQuery.addDomEvent(scope.getWin(), "resize", updateSize)
        ];
        return () => {
          cleanups.forEach((cleanup) => cleanup?.());
        };
      },
      trackNestedDrawerMetrics({ scope, prop }) {
        const contentEl = getContentEl(scope);
        if (!contentEl) return;
        const id = String(prop("id") ?? scope.id);
        drawerRegistry.register(id, contentEl);
        const sync = () => {
          const entries = [...drawerRegistry.getEntries().entries()];
          const myIndex = entries.findIndex(([entryId]) => entryId === id);
          if (myIndex === -1) return;
          const nestedCount = entries.length - 1 - myIndex;
          const frontmostEl = entries[entries.length - 1]?.[1];
          const frontmostHeight = frontmostEl?.getBoundingClientRect().height ?? 0;
          const myHeight = contentEl.getBoundingClientRect().height;
          contentEl.style.setProperty("--nested-drawers", `${nestedCount}`);
          contentEl.style.setProperty("--drawer-height", `${myHeight}px`);
          contentEl.style.setProperty("--drawer-frontmost-height", `${frontmostHeight}px`);
          if (nestedCount > 0) contentEl.setAttribute("data-nested-drawer-open", "");
          else contentEl.removeAttribute("data-nested-drawer-open");
          if (drawerRegistry.hasSwipingAfter(id)) contentEl.setAttribute("data-nested-drawer-swiping", "");
          else contentEl.removeAttribute("data-nested-drawer-swiping");
        };
        sync();
        const cleanups = [
          drawerRegistry.subscribe(sync),
          domQuery.resizeObserverBorderBox.observe(contentEl, () => drawerRegistry.notify()),
          domQuery.addDomEvent(scope.getWin(), "resize", () => drawerRegistry.notify())
        ];
        return () => {
          cleanups.forEach((cleanup) => cleanup?.());
          contentEl.removeAttribute("data-nested-drawer-open");
          contentEl.removeAttribute("data-nested-drawer-swiping");
          contentEl.style.setProperty("--nested-drawers", "0");
          contentEl.style.removeProperty("--drawer-frontmost-height");
          drawerRegistry.unregister(id);
        };
      },
      trackExitAnimation({ send, scope }) {
        const contentEl = getContentEl(scope);
        if (!contentEl) return;
        return domQuery.addDomEvent(contentEl, "exitcomplete", () => {
          send({ type: "ANIMATION_END" });
        });
      }
    }
  }
});
var props = types.createProps()([
  "id",
  "ids",
  "dir",
  "modal",
  "initialFocusEl",
  "finalFocusEl",
  "open",
  "defaultOpen",
  "getRootNode",
  "snapPoints",
  "swipeDirection",
  "snapToSequentialPoints",
  "swipeVelocityThreshold",
  "closeThreshold",
  "preventDragOnScroll",
  "closeOnEscape",
  "closeOnInteractOutside",
  "onEscapeKeyDown",
  "onFocusOutside",
  "onInteractOutside",
  "onOpenChange",
  "onPointerDownOutside",
  "onRequestDismiss",
  "preventScroll",
  "restoreFocus",
  "role",
  "trapFocus",
  "defaultSnapPoint",
  "snapPoint",
  "onSnapPointChange",
  "stack"
]);
var splitProps = utils.createSplitProps(props);

// src/drawer.stack.ts
function clamp3(value, min, max) {
  return Math.min(max, Math.max(min, value));
}
function resolveSnapshot(entries) {
  let openCount = 0;
  let frontmostHeight = 0;
  let swipeProgress = 0;
  let frontmostOrder = -1;
  entries.forEach((entry) => {
    if (!entry.open) return;
    openCount += 1;
    if (entry.order < frontmostOrder) return;
    frontmostOrder = entry.order;
    frontmostHeight = entry.height > 0 ? entry.height : 0;
    swipeProgress = entry.swiping ? clamp3(entry.swipeProgress, 0, 1) : 0;
  });
  return {
    active: openCount > 0,
    openCount,
    swipeProgress,
    frontmostHeight
  };
}
function createStack() {
  const entries = /* @__PURE__ */ new Map();
  const listeners2 = /* @__PURE__ */ new Set();
  let order = 0;
  let pendingNotify = false;
  let snapshot = Object.freeze({
    active: false,
    openCount: 0,
    swipeProgress: 0,
    frontmostHeight: 0
  });
  const scheduleNotify = () => {
    if (pendingNotify) return;
    pendingNotify = true;
    queueMicrotask(() => {
      pendingNotify = false;
      listeners2.forEach((listener) => listener());
    });
  };
  const notify2 = () => {
    const nextSnapshot = resolveSnapshot(entries);
    if (snapshot.active === nextSnapshot.active && snapshot.openCount === nextSnapshot.openCount && snapshot.swipeProgress === nextSnapshot.swipeProgress && snapshot.frontmostHeight === nextSnapshot.frontmostHeight) {
      return;
    }
    snapshot = Object.freeze(nextSnapshot);
    scheduleNotify();
  };
  const ensureEntry = (id) => {
    let entry = entries.get(id);
    if (!entry) {
      entry = { order: ++order, open: false, height: 0, swiping: false, swipeProgress: 0 };
      entries.set(id, entry);
    }
    return entry;
  };
  return {
    getSnapshot() {
      return snapshot;
    },
    subscribe(listener) {
      listeners2.add(listener);
      return () => {
        listeners2.delete(listener);
      };
    },
    register(id) {
      ensureEntry(id);
      notify2();
    },
    unregister(id) {
      if (!entries.delete(id)) return;
      notify2();
    },
    setOpen(id, open) {
      const entry = ensureEntry(id);
      if (entry.open === open) return;
      entry.open = open;
      if (!open) {
        entry.swiping = false;
        entry.swipeProgress = 0;
      }
      notify2();
    },
    setHeight(id, height) {
      const entry = ensureEntry(id);
      const nextHeight = Number.isFinite(height) && height > 0 ? height : 0;
      if (entry.height === nextHeight) return;
      entry.height = nextHeight;
      notify2();
    },
    setSwipe(id, swiping, progress) {
      const entry = ensureEntry(id);
      const nextProgress = swiping ? clamp3(Number.isFinite(progress) ? progress : 0, 0, 1) : 0;
      if (entry.swiping === swiping && entry.swipeProgress === nextProgress) return;
      entry.swiping = swiping;
      entry.swipeProgress = nextProgress;
      notify2();
    }
  };
}
function connectStack(snapshot, normalize) {
  const getIndentProps = () => {
    return normalize.element({
      "data-active": snapshot.active ? "" : void 0,
      "data-inactive": snapshot.active ? void 0 : "",
      style: {
        "--drawer-swipe-progress": `${clamp3(snapshot.swipeProgress, 0, 1)}`,
        "--drawer-frontmost-height": snapshot.frontmostHeight > 0 ? `${snapshot.frontmostHeight}px` : void 0
      }
    });
  };
  return {
    getIndentProps,
    getIndentBackgroundProps: getIndentProps
  };
}

exports.anatomy = anatomy;
exports.connect = connect;
exports.connectStack = connectStack;
exports.createStack = createStack;
exports.machine = machine;
exports.props = props;
exports.splitProps = splitProps;
